### ISSUE TITLE
rustup update causes clion debugging to fail

### ISSUE THREAD

**Issue by nixomose** [rust-lang/rust#72196]
A guy I was working with told me that there were new rustfmt features available in stable, and it was my task to reformat the code in our project so I did rustup update and it updated from 1.40.0 which is what I had to 1.43.1
the fmt features weren't there "fn_params_layout" doesn't work apparently in 1.40.0 it's called "fn_args_layout" but what we were looking for was "fn_call_width" which according to this:  https://rust-lang.github.io/rustfmt/ is an option, but it doesn't work.

So anyway, I gave up and went back to debugging a problem I was working on.
I use clion with the latest rust plugin, but now when I try to run in the debugger, I get 

```
/home/smark/git/xxx/target/debug/yyy
/home/smark/git/xxx/target/debug/yyy: relocation error: /home/smark/git/xxx/target/debug/yyy: symbol �tH�H;������H���f.� version OPENSSL_1_1_0 not defined in file libcrypto.so.1.1 with link time referenceProcess finished with exit code 127
```

if I run the program from the command line it works, and if I run it from gdb, it works, it only fails, when I try and run it from the debugger in clion.

There are two other debugging options gdb in /usr/bin and lldb.
so I tried them both.
my local gdb yielded the same result as the bundled gdb which was the default option.

but lldb gave me a more interesting error.

```
/home/smark/git/xxx/target/debug/yyy
Inconsistency detected by ld.so: rtld.c: 1266: dl_main: Assertion `GL(dl_rtld_map).l_libname' failed!
```

That didn't help me any, but it convinced me that something was wrong with the build.
Although again it still runs from the command line, but then I figured the environment the debugger runs in must be different than the environment that my shell has.
I couldn't find too much in the way of configurability with the debugger in clion, there's a place to set environment variables, so I tried setting LD_LIBRARY_PATH to where the libssl libraries are that the first error message was complaining about, but that didn't help.

Then I figured maybe I'd get rid of libssl 1.1 and just use 1.0 and after shuffling the packages, no change same error.

I googled a bunch, nobody had this problem with rust but me. There are other reports of that problem with other packages and the solution always seems to be rebuild with this or that switch.
I can't really rebuild rust, it's not my job. I'm just trying to debug my rust program.

Then what happened, trying to remember, it was a long day yesterday first with the formatting options then the debugger problem.
Oh yeah, so the formatter thing, I tried to set width_heuristics, but it wanted  a structure. after a bunch of trial and error I was able to create the structure it wanted in the rustfmt.toml file and then when I finally got it all right it said something like "this is a nightly feature only". so that was a dead end too.

But again, I gave up on that formatting thing, back to the debugger problem.

So after a while it finally occurred to me that the only other thing I did was the rustup update, maybe that had something to do with it.

So I tried to figure out what version of rustup I had before updating. if it printed it out, I didn't see it or recall it and that shell was gone.

But I run zfs for my filesystem, and I have it take a snapshot of my filesystem every minute so I can go back in time and see what happened.
So I sifted back to the beginning of the day when I was able to use the debugger and tried to figure out what version of cargo and rustc and rustup I had.
I found the binaries in .cargo/bin, but while they're all the same size, and copies of the same file, they're not actually hardlinks. I didn't quite understand that. But whatever.
so I went to my early morning snapshot ran ./rustup --version and it said 1.43.1 which was the latest that rustup had updated to. I thought that odd, so I went back further in time and kept getting the same thing.

it seems that something about the binary when it loads must look in the environment and run or link in the current binary because it kept getting me the current version not the one I was actually running.
maybe that's a weird side effect of rust and zfs snapshots, I dunno.

Eventually I got the idea to just run strings on the binary, and then I was able to tell that the version of rustup I had that morning was 1.40.0

so I ran rustup default 1.40.0, rebuilt everything and then my debugger worked.

I told my boss this story and he said 

"I'd file a bug w/them, and describe what you're seeing... curious to see how quickly they respond to this"

So here I am.

Another glaring piece of information I came across was this:
https://github.com/rust-lang/rust/blob/master/RELEASES.md
the first very line says 

```
"Updated openssl-src to 1.1.1g for CVE-2020-1967."
```

boy if that doesn't look incriminating.

I had tried to downgrade to 1.43.0 but that didn't work either, but I admit I might have messed up the install. this event had been going on for quite a while at this point.

I figured 1.40.0 had worked that morning, I'd go back to what I had, and that did, indeed fix it.

So there's my bug report.

I expected to see this happen: *explanation*

Instead, this happened: *explanation*

### Meta

I found a workaround, that's good enough for now.

`rustc --version --verbose`:
```
rustc 1.40.0 (73528e339 2019-12-16)
binary: rustc
commit-hash: 73528e339aae0f17a15ffa49a8ac608f50c6cf14
commit-date: 2019-12-16
host: x86_64-unknown-linux-gnu
release: 1.40.0
LLVM version: 9.0

```

<details><summary>Backtrace</summary>
<p>

```
there is no backtrace, there is only the error message above, I'll repeat it here:

/home/smark/git/xxx/target/debug/yyy: relocation error: /home/smark/git/xxx/target/debug/yyy: symbol �tH�H;������H���f.� version OPENSSL_1_1_0 not defined in file libcrypto.so.1.1 with link time referenceProcess finished with exit code 127



```

</p>
</details>

**Comment by maxammann** [comment]
I have the same problem, sad that nobody else has this one :cry: 

I get in GDB in CLion:
```
Inconsistency detected by ld.so: dl-version.c: 205: _dl_check_map_versions: Assertion `needed != NULL' failed!
```

It works perfectly when running gdb on my system and not through CLion. I already switched JDKs which did not make difference. It also is not a difference if I use the system or bundled GDB in CLion.

**Comment by maxammann** [comment]
Alright I found a workaround for me:

I had quite many breakpoints set in CLion for external crates. Now If I disable all this works. This is probably the reason why the behavior between GDB in CLion and in a Termal was different. If you have breakpoints in Clion then they are set in GDB. 

Now if I run GDB with these breakpoints it works:

![image](https://user-images.githubusercontent.com/905221/115010262-7dadb900-9ead-11eb-8374-cf163d61e931.png)

If I set this these breakpoints I get a symbol lookup error:
```
symbol lookup error: xxxx: undefined symbol: lock, version GLIBC_2.2.5
```

![image](https://user-images.githubusercontent.com/905221/115010406-afbf1b00-9ead-11eb-955f-c91b5fc85bef.png)

If I set only the first one in the second screenshot I get a SEGV:

![image](https://user-images.githubusercontent.com/905221/115010489-c7969f00-9ead-11eb-8264-5bf3459f493b.png)
![image](https://user-images.githubusercontent.com/905221/115010533-d2513400-9ead-11eb-9508-bc1d563dd515.png)


If I set these breakpoints I get this error:
```
Inconsistency detected by ld.so: dl-version.c: 205: _dl_check_map_versions: Assertion `needed != NULL' failed!
```
![image](https://user-images.githubusercontent.com/905221/115010718-0593c300-9eae-11eb-8496-162535b283d4.png)


This is the crate which I'm debugging:
https://github.com/ctz/rustls/tree/v/0.19.0

**Comment by nixomose** [comment]
hey so I found a workaround.

it seems to be a problem with the linker, and the default optimizations that are on in debug mode.
I have no idea why any optimizations are on when you're compiling and linking for debug builds, but they are and they cause the problem.
So if you turn them off, the problem goes away...
Here's what I did.

since I didn't want to commit this change to git, I wrote a little script to make the change for me...


#!/bin/bash

# now check to see if the profile.dev stuff is in root cargo.toml and if not, add it
grep profile.dev /home/smark/git/dblock/Cargo.toml
if [ "$?" != 0 ]; then
  echo "profile.dev not found in cargo.toml, adding it."
  echo "# avoid any optimization, so that the linker might not destroy the binary and make it undebuggable. anger. hate." >> /home/nixomose/git/project/Cargo.toml
  echo '[profile.dev]' >> /home/smark/git/dblock/Cargo.toml
  echo "incremental = false" >> /home/smark/git/dblock/Cargo.toml
  echo "codegen-units = 1" >> /home/smark/git/dblock/Cargo.toml
  echo 'lto = "off"' >> /home/smark/git/dblock/Cargo.toml
fi

in short you need to add this to your cargo.toml

[profile.dev]
incremental = false
codegen-units = 1
lto = "off"

I don't know which one causes the problem, but setting them all makes it go away.

nixomose: 1
crappy software: 0

**Comment by maxammann** [comment]
Thanks for the hints! Maybe this helps the rust developers to inspect this issue. So by remoing the old breakpoints I can debug now again. So this also works without the flags for me.

nixomose: 1
max: 1
crappy software: 0

**Comment by Enselic** [comment]
Triage: Is this still a problem? If so, please provide as simple step by step instructions on how to reproduce as possible.

**Comment by nixomose** [comment]
well I haven't done rust development in a while so I don't know, but I'm going to guess probably not fixed. it wasn't easy to reproduce, but I was able to show that it was some optimization causing the problem. 
adding this to your cargo.toml makes it start working again

[profile.dev]
incremental = false
codegen-units = 1
lto = "off"

**Comment by Noratrieb** [comment]
an old issue without a reproduction, closing. if someone has this problem again, open a new issue.
